cmd/secrets/secret_helper.go

```diff
@@ -53,105 +81,112 @@ type secretsRequest struct {
 	}
 
 	if len(request.Secrets) == 0 {
-		return errors.New("no secrets listed in input")
+		return nil, errors.New("no secrets listed in input")
 	}
 
-	response := map[string]s.Secret{}
-	for _, name := range request.Secrets {
-		response[name] = readSecret(filepath.Join(dir, name))
-	}
+	return request.Secrets, nil
+}
 
-	out, err := json.Marshal(response)
+func writeDecodedSecrets(w io.Writer, resolvedSecrets map[string]s.Secret) error {
+	out, err := json.Marshal(resolvedSecrets)
 	if err != nil {
 		return err
 	}
+
 	_, err = w.Write(out)
 	return err
 }
 
-func splitVersion(ver string) []string {
-	return strings.SplitN(ver, ".", 2)
-}
+func readSecretsFromFile(secrets []string, dir string) map[string]s.Secret {
+	res := make(map[string]s.Secret)
 
-func readSecret(path string) s.Secret {
-	value, err := readSecretFile(path)
-	var errMsg string
-	if err != nil {
-		errMsg = err.Error()
+	secretBackend := fileBackend{rootPath: dir}
+	for _, secretID := range secrets {
+		res[secretID] = secretBackend.get(secretID)
 	}
-	return s.Secret{Value: value, ErrorMsg: errMsg}
+
+	return res
 }
 
-func readSecretFile(path string) (string, error) {
-	fi, err := os.Lstat(path)
-	if err != nil {
-		if os.IsNotExist(err) {
-			return "", errors.New("secret does not exist")
-		}
-		return "", err
-	}
+func readSecretsUsingPrefixes(secrets []string, selector *backendSelector) map[string]s.Secret {
+	res := make(map[string]s.Secret)
 
-	if fi.Mode()&os.ModeSymlink != 0 {
-		// Ensure that the symlink is in the same dir
-		target, err := os.Readlink(path)
+	for _, secretID := range secrets {
+		prefix, id, err := parseSecretWithPrefix(secretID)
 		if err != nil {
-			return "", fmt.Errorf("failed to read symlink target: %v", err)
-		}
-
-		dir := filepath.Dir(path)
-		if !filepath.IsAbs(target) {
-			target, err = filepath.Abs(filepath.Join(dir, target))
-			if err != nil {
-				return "", fmt.Errorf("failed to resolve symlink absolute path: %v", err)
-			}
+			res[secretID] = s.Secret{Value: "", ErrorMsg: err.Error()}
+			continue
 		}
 
-		dirAbs, err := filepath.Abs(dir)
+		secretBackend, err := selector.choose(prefix)
 		if err != nil {
-			return "", fmt.Errorf("failed to resolve absolute path of directory: %v", err)
+			res[secretID] = s.Secret{Value: "", ErrorMsg: err.Error()}
+			continue
 		}
 
-		if !filepath.HasPrefix(target, dirAbs) {
-			return "", fmt.Errorf("not following symlink %q outside of %q", target, dir)
-		}
-	}
-	fi, err = os.Stat(path)
-	if err != nil {
-		return "", err
+		res[secretID] = secretBackend.get(id)
 	}
 
-	if fi.Size() > maxSecretFileSize {
-		return "", errors.New("secret exceeds max allowed size")
-	}
+	return res
+}
 
-	file, err := os.Open(path)
-	if err != nil {
-		return "", err
+func parseSecretWithPrefix(secretID string) (prefix backendPrefix, id string, err error) {
+	split := strings.SplitN(secretID, backendPrefixSeparator, 2)
+	if len(split) != 2 {
```

Comment:
nit
```suggestion
	if len(split) < 2 {
```

Reply:
You're right. We're calling `splitN` with 2, so we know that it cannot be > 2.
Changed :+1: